packages/go/go-helpers.ts

```diff
@@ -112,72 +112,146 @@ class GoWrapper {
-  }
+  const env = cloneEnv(process.env, opts.env);
+  const { arch, platform } = process;
+  const goGlobalDir = join(
+    goGlobalCachePath,
+    `${goSelectedVersion}_${platform}_${arch}`
+  );
+  const goGlobalBinDir = join(goGlobalDir, 'bin');
 
-  // Check `go` bin in cacheDir
-  const isGoExist = await pathExists(join(dir, 'bin'));
-  if (!isGoExist) {
-    debug('Installing `go` v%s to %o for %s %s', version, dir, platform, arch);
-    const url = getGoUrl(version, platform, arch);
-    debug('Downloading `go` URL: %o', url);
-    const res = await fetch(url);
+  if (goPreferredVersion) {
+    debug(`Preferred go version ${goPreferredVersion} (from go.mod)`);
+    env.GO111MODULE = 'on';
+  }
 
-    if (!res.ok) {
-      throw new Error(`Failed to download: ${url} (${res.status})`);
+  // check we have the desired `go` version cached
+  if (await pathExists(goGlobalBinDir)) {
+    // check if `go` has already been downloaded and that the version is correct
+    env.GOROOT = goGlobalDir;
+    env.PATH = `${goGlobalBinDir}${delimiter}${env.PATH}`;
+    const { failed, stdout } = await execa('go', ['version'], {
+      env,
+      reject: false,
+    });
+    if (!failed) {
+      const { version, short } = parseGoVersionString(stdout);
+      if (version === goSelectedVersion || short === goSelectedVersion) {
+        debug(`Selected go ${version} (from cache)`);
+        return new GoWrapper(env, opts);
+      } else {
+        debug(`Found cached go ${version}, but need ${goSelectedVersion}`);
+      }
     }
+  }
 
-    // TODO: use a zip extractor when `ext === "zip"`
-    await mkdirp(dir);
-    await new Promise((resolve, reject) => {
-      res.body
-        .on('error', reject)
-        .pipe(tar.extract({ cwd: dir, strip: 1 }))
-        .on('error', reject)
-        .on('finish', resolve);
+  if (!goPreferredVersion) {
+    // check if `go` is installed in the system PATH and if it's the version we want
+    const { failed, stdout } = await execa('go', ['version'], {
+      env,
+      reject: false,
     });
+    if (!failed) {
+      const { version, minor } = parseGoVersionString(stdout);
+      if (minor < GO_MIN_VERSION) {
+        debug(`Found go ${version} in system PATH, but version is unsupported`);
+      } else if (!goPreferredVersion || goPreferredVersion === version) {
+        debug(`Selected go ${version} (from system PATH)`);
+        return new GoWrapper(env, opts);
+      } else {
+        debug(
+          `Found go ${version} in system PATH, but preferred version is ${goPreferredVersion}`
+        );
+      }
+    }
   }
-  return createGo(workPath, dir, platform, arch);
+
+  // we need to download and cache the desired `go` version
+  const url = getGoUrl(goSelectedVersion, platform, arch);
+  debug(`Downloading go: ${url}`);
+  const res = await fetch(url);
+
+  if (!res.ok) {
+    throw new Error(`Failed to download: ${url} (${res.status})`);
+  }
+
+  // TODO: use a zip extractor when `ext === "zip"`
```

Comment:
**question (non-blocking):** when will this be a `.zip`?

Reply:
Turns out that all Go distributions for Windows are `.zip` files and it apparently has never been supported. I added the ability to extract zip files and now Windows is happy.